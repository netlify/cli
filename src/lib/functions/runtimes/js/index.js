const { dirname } = require('path')

const lambdaLocal = require('lambda-local')
const winston = require('winston')

const detectNetlifyLambdaBuilder = require('./builders/netlify-lambda')
const detectZisiBuilder = require('./builders/zisi')

const SECONDS_TO_MILLISECONDS = 1e3

let netlifyLambdaDetectorCache

const logger = winston.createLogger({
  levels: winston.config.npm.levels,
  transports: [new winston.transports.Console({ level: 'warn' })],
})

lambdaLocal.setLogger(logger)

// The netlify-lambda builder can't be enabled or disabled on a per-function
// basis and its detection mechanism is also quite expensive, so we detect
// it once and cache the result.
const detectNetlifyLambdaWithCache = () => {
  if (netlifyLambdaDetectorCache === undefined) {
    netlifyLambdaDetectorCache = detectNetlifyLambdaBuilder()
  }

  return netlifyLambdaDetectorCache
}

const getBuildFunction = async ({ config, directory, errorExit, func, projectRoot }) => {
  const netlifyLambdaBuilder = await detectNetlifyLambdaWithCache()

  if (netlifyLambdaBuilder) {
    return netlifyLambdaBuilder.build
  }

  const zisiBuilder = await detectZisiBuilder({ config, directory, errorExit, func, projectRoot })

  if (zisiBuilder) {
    return zisiBuilder.build
  }

  // If there's no function builder, we create a simple one on-the-fly which
  // returns as `srcFiles` the function directory, if there is one, or its
  // main file otherwise.
  const functionDirectory = dirname(func.mainFile)
  const srcFiles = functionDirectory === directory ? [func.mainFile] : [functionDirectory]
  const schedule = await detectZisiBuilder.parseForSchedule({ mainFile: func.mainFile, config, projectRoot })

  return () => ({ schedule, srcFiles })
}

const invokeFunction = async ({ context, event, func, timeout }) => {
  // If a function builder has defined a `buildPath` property, we use it.
  // Otherwise, we'll invoke the function's main file.
  const lambdaPath = (func.buildData && func.buildData.buildPath) || func.mainFile
  const result = await lambdaLocal.execute({
    clientContext: JSON.stringify(context),
    event,
    lambdaPath,
    timeoutMs: timeout * SECONDS_TO_MILLISECONDS,
    verboseLevel: 3,
  })

  return result
}

const onDirectoryScan = async () => {
  const netlifyLambdaBuilder = await detectNetlifyLambdaWithCache()

  // Before we start a directory scan, we check whether netlify-lambda is being
  // used. If it is, we run it, so that the functions directory is populated
  // with the compiled files before the scan begins.
  if (netlifyLambdaBuilder) {
    await netlifyLambdaBuilder.build()
  }
}

module.exports = { getBuildFunction, invokeFunction, name: 'js', onDirectoryScan }
