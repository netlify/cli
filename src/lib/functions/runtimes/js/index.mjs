import { dirname } from 'path'

import lambdaLocal from '@skn0tt/lambda-local'
import winston from 'winston'

import detectNetlifyLambdaBuilder from './builders/netlify-lambda.mjs'
import detectZisiBuilder, { parseForSchedule } from './builders/zisi.mjs'

export const name = 'js'

const SECONDS_TO_MILLISECONDS = 1000

let netlifyLambdaDetectorCache

const logger = winston.createLogger({
  levels: winston.config.npm.levels,
  transports: [new winston.transports.Console({ level: 'warn' })],
})

lambdaLocal.setLogger(logger)

// The netlify-lambda builder can't be enabled or disabled on a per-function
// basis and its detection mechanism is also quite expensive, so we detect
// it once and cache the result.
const detectNetlifyLambdaWithCache = () => {
  if (netlifyLambdaDetectorCache === undefined) {
    netlifyLambdaDetectorCache = detectNetlifyLambdaBuilder()
  }

  return netlifyLambdaDetectorCache
}

export const getBuildFunction = async ({ config, directory, errorExit, func, projectRoot }) => {
  const netlifyLambdaBuilder = await detectNetlifyLambdaWithCache()

  if (netlifyLambdaBuilder) {
    return netlifyLambdaBuilder.build
  }

  const zisiBuilder = await detectZisiBuilder({ config, directory, errorExit, func, projectRoot })

  if (zisiBuilder) {
    return zisiBuilder.build
  }

  // If there's no function builder, we create a simple one on-the-fly which
  // returns as `srcFiles` the function directory, if there is one, or its
  // main file otherwise.
  const functionDirectory = dirname(func.mainFile)
  const srcFiles = functionDirectory === directory ? [func.mainFile] : [functionDirectory]
  const schedule = await parseForSchedule({ mainFile: func.mainFile, config, projectRoot })

  return () => ({ schedule, srcFiles })
}

export const invokeFunction = async ({ context, event, func, timeout }) => {
  // If a function builder has defined a `buildPath` property, we use it.
  // Otherwise, we'll invoke the function's main file.
  const lambdaPath = (func.buildData && func.buildData.buildPath) || func.mainFile
  const result = await lambdaLocal.execute({
    clientContext: JSON.stringify(context),
    event,
    lambdaPath,
    timeoutMs: timeout * SECONDS_TO_MILLISECONDS,
    verboseLevel: 3,
  })

  return result
}

export const onDirectoryScan = async () => {
  const netlifyLambdaBuilder = await detectNetlifyLambdaWithCache()

  // Before we start a directory scan, we check whether netlify-lambda is being
  // used. If it is, we run it, so that the functions directory is populated
  // with the compiled files before the scan begins.
  if (netlifyLambdaBuilder) {
    await netlifyLambdaBuilder.build()
  }
}
