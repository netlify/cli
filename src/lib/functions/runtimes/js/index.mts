// @ts-expect-error TS(2451): Cannot redeclare block-scoped variable 'dirname'.
const { dirname } = require('path')

const lambdaLocal = require('lambda-local')
const winston = require('winston')

const detectNetlifyLambdaBuilder = require('./builders/netlify-lambda.cjs')
const detectZisiBuilder = require('./builders/zisi.cjs')

// @ts-expect-error TS(2451): Cannot redeclare block-scoped variable 'SECONDS_TO... Remove this comment to see the full error message
const SECONDS_TO_MILLISECONDS = 1e3

// @ts-expect-error TS(2304): Cannot find name '$TSFixMe'.
let netlifyLambdaDetectorCache: $TSFixMe

const logger = winston.createLogger({
  levels: winston.config.npm.levels,
  transports: [new winston.transports.Console({ level: 'warn' })],
})

lambdaLocal.setLogger(logger)

// The netlify-lambda builder can't be enabled or disabled on a per-function
// basis and its detection mechanism is also quite expensive, so we detect
// it once and cache the result.
const detectNetlifyLambdaWithCache = () => {
  if (netlifyLambdaDetectorCache === undefined) {
    netlifyLambdaDetectorCache = detectNetlifyLambdaBuilder()
  }

  return netlifyLambdaDetectorCache
}

// @ts-expect-error TS(2451): Cannot redeclare block-scoped variable 'getBuildFu... Remove this comment to see the full error message
const getBuildFunction = async ({
  config,
  directory,
  errorExit,
  func,
  projectRoot
// @ts-expect-error TS(2304): Cannot find name '$TSFixMe'.
}: $TSFixMe) => {
  const netlifyLambdaBuilder = await detectNetlifyLambdaWithCache()

  if (netlifyLambdaBuilder) {
    return netlifyLambdaBuilder.build
  }

  const zisiBuilder = await detectZisiBuilder({ config, directory, errorExit, func, projectRoot })

  if (zisiBuilder) {
    return zisiBuilder.build
  }

  // If there's no function builder, we create a simple one on-the-fly which
  // returns as `srcFiles` the function directory, if there is one, or its
  // main file otherwise.
  const functionDirectory = dirname(func.mainFile)
  const srcFiles = functionDirectory === directory ? [func.mainFile] : [functionDirectory]
  const schedule = await detectZisiBuilder.parseForSchedule({ mainFile: func.mainFile, config, projectRoot })

  return () => ({ schedule, srcFiles })
}

// @ts-expect-error TS(2451): Cannot redeclare block-scoped variable 'invokeFunc... Remove this comment to see the full error message
const invokeFunction = async ({
  context,
  event,
  func,
  timeout
// @ts-expect-error TS(2304): Cannot find name '$TSFixMe'.
}: $TSFixMe) => {
  // If a function builder has defined a `buildPath` property, we use it.
  // Otherwise, we'll invoke the function's main file.
  const lambdaPath = (func.buildData && func.buildData.buildPath) || func.mainFile
  const result = await lambdaLocal.execute({
    clientContext: JSON.stringify(context),
    event,
    lambdaPath,
    timeoutMs: timeout * SECONDS_TO_MILLISECONDS,
    verboseLevel: 3,
  })

  return result
}

const onDirectoryScan = async () => {
  const netlifyLambdaBuilder = await detectNetlifyLambdaWithCache()

  // Before we start a directory scan, we check whether netlify-lambda is being
  // used. If it is, we run it, so that the functions directory is populated
  // with the compiled files before the scan begins.
  if (netlifyLambdaBuilder) {
    await netlifyLambdaBuilder.build()
  }
}

module.exports = { getBuildFunction, invokeFunction, name: 'js', onDirectoryScan }
