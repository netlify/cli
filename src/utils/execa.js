const { env } = require('process')

// This is a thin layer on top of `execa` that allows consumers to provide an
// alternative path to the module location, making it easier to mock its logic
// in tests (see `tests/utils/mock-execa.js`).
// eslint-disable-next-line import/no-dynamic-require
const execa = require(env.NETLIFY_CLI_EXECA_PATH || 'execa')

module.exports = execa
