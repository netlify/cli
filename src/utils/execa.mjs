import { env } from 'process'
// This is a thin layer on top of `execa` that allows consumers to provide an
// alternative path to the module location, making it easier to mock its logic
// in tests (see `tests/utils/mock-execa.js`).
// eslint-disable-next-line import/no-mutable-exports
let execa

if (env.NETLIFY_CLI_EXECA_PATH) {
  // eslint-disable-next-line import/no-dynamic-require
  const execaMock = await import(env.NETLIFY_CLI_EXECA_PATH)
  execa = execaMock.default
} else {
  const execaLib = await import('execa')
  execa = execaLib.default
}

export default execa
