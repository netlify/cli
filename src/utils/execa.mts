const { env } = require('process')

const execaLib = require('execa')

// This is a thin layer on top of `execa` that allows consumers to provide an
// alternative path to the module location, making it easier to mock its logic
// in tests (see `tests/utils/mock-execa.js`).

// eslint-disable-next-line import/no-dynamic-require
const execa = env.NETLIFY_CLI_EXECA_PATH ? require(env.NETLIFY_CLI_EXECA_PATH) : execaLib

module.exports = execa
