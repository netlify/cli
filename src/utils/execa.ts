import { env } from 'process'

import execaLib from 'execa'

// This is a thin layer on top of `execa` that allows consumers to provide an
// alternative path to the module location, making it easier to mock its logic
// in tests (see `tests/utils/mock-execa.js`).

/**
 * @type {import('execa')}
 */
// eslint-disable-next-line import/no-mutable-exports
let execa

if (env.NETLIFY_CLI_EXECA_PATH) {
  const execaMock = await import(env.NETLIFY_CLI_EXECA_PATH)
  execa = execaMock.default
} else {
  execa = execaLib
}

export default execa
