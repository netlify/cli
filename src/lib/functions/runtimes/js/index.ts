import { createConnection } from 'net'
import { dirname } from 'path'
import { pathToFileURL } from 'url'
import { Worker } from 'worker_threads'

import lambdaLocal, { type LambdaEvent } from 'lambda-local'

import type { BuildFunction, GetBuildFunction, InvokeFunction } from '../index.js'
import { BLOBS_CONTEXT_VARIABLE } from '../../../blobs/blobs.js'
import type NetlifyFunction from '../../netlify-function.js'

import detectZisiBuilder, { getFunctionMetadata, ZisiBuildResult } from './builders/zisi.js'
import { SECONDS_TO_MILLISECONDS } from './constants.js'
import type { WorkerMessage } from './worker.js'

export const name = 'js'

type SimpleJsBuildResult = {
  schedule?: string
  srcFiles: string[]
}

export type JsBuildResult = ZisiBuildResult | SimpleJsBuildResult

// TODO(serhalp): Unify these. This is bonkers that the two underlying invocation mechanisms are encapsulated but we
// return slightly different shapes for them.
export type JsInvokeFunctionResult = WorkerMessage | LambdaEvent

lambdaLocal.getLogger().level = 'alert'

export async function getBuildFunction({
  config,
  directory,
  errorExit,
  func,
  projectRoot,
}: Parameters<GetBuildFunction<JsBuildResult>>[0]) {
  const metadata = await getFunctionMetadata({ mainFile: func.mainFile, config, projectRoot })
  const zisiBuilder = await detectZisiBuilder({ config, directory, errorExit, func, metadata, projectRoot })

  if (zisiBuilder) {
    return zisiBuilder.build
  }

  // If there's no function builder, we create a simple one on-the-fly which
  // returns as `srcFiles` the function directory, if there is one, or its
  // main file otherwise.
  const functionDirectory = dirname(func.mainFile)
  const srcFiles = functionDirectory === directory ? [func.mainFile] : [functionDirectory]

  const build: BuildFunction<JsBuildResult> = () => Promise.resolve({ schedule: metadata?.schedule, srcFiles })
  return build
}

const workerURL = new URL('worker.js', import.meta.url)

export const invokeFunction = async ({
  context,
  environment,
  event,
  func,
  timeout,
}: Parameters<InvokeFunction<JsBuildResult>>[0]): Promise<JsInvokeFunctionResult> => {
  const { buildData } = func
  // I have no idea why, but it appears that treating the case of a missing `buildData` or missing
  // `buildData.runtimeAPIVersion` as V1 is important.
  const runtimeAPIVersion =
    buildData != null && 'runtimeAPIVersion' in buildData && typeof buildData.runtimeAPIVersion === 'number'
      ? buildData.runtimeAPIVersion
      : null
  if (runtimeAPIVersion == null || runtimeAPIVersion !== 2) {
    return await invokeFunctionDirectly({
      context,
      environment: environment as Record<string, string>,
      event,
      func,
      timeout,
    })
  }

  const workerData = {
    clientContext: JSON.stringify(context),
    environment,
    event,
    // If a function builder has defined a `buildPath` property, we use it.
    // Otherwise, we'll invoke the function's main file.
    // Because we use import() we have to use file:// URLs for Windows.
    entryFilePath: pathToFileURL(
      buildData != null && 'buildPath' in buildData && buildData.buildPath ? buildData.buildPath : func.mainFile,
    ).href,
    timeoutMs: timeout * SECONDS_TO_MILLISECONDS,
  }

  const worker = new Worker(workerURL, { workerData })
  return await new Promise((resolve, reject) => {
    worker.on('message', (result: WorkerMessage): void => {
      // TODO(serhalp): Improve `WorkerMessage` type. It sure would be nice to keep it simple as it
      // is now, but technically this is an arbitrary type from the user function return...
      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
      if (result?.streamPort != null) {
        const client = createConnection(
          {
            port: result.streamPort,
            host: 'localhost',
          },
          () => {
            result.body = client
            resolve(result)
          },
        )
        client.on('error', reject)
      } else {
        resolve(result)
      }
    })

    worker.on('error', reject)
  })
}

export const invokeFunctionDirectly = async <BuildResult extends JsBuildResult>({
  context,
  environment,
  event,
  func,
  timeout,
}: {
  context: Record<string, unknown>
  environment: Record<string, string>
  event: Record<string, unknown>
  func: NetlifyFunction<BuildResult>
  timeout: number
}): Promise<LambdaEvent> => {
  const buildData = await func.getBuildData()
  if (buildData == null) {
    throw new Error('Cannot invoke a function that has not been built')
  }
  // If a function builder has defined a `buildPath` property, we use it.
  // Otherwise, we'll invoke the function's main file.
  const lambdaPath =
    'buildPath' in buildData && typeof buildData.buildPath === 'string' ? buildData.buildPath : func.mainFile
  const result = await lambdaLocal.execute({
    clientContext: JSON.stringify(context),
    environment: {
      // Include environment variables from config
      ...environment,
      // We've set the Blobs context on the parent process, which means it will
      // be available to the Lambda. This would be inconsistent with production
      // where only V2 functions get the context injected. To fix it, unset the
      // context variable before invoking the function.
      // This has the side-effect of also removing the variable from `process.env`.
      [BLOBS_CONTEXT_VARIABLE]: undefined,
    },
    event,
    lambdaPath,
    timeoutMs: timeout * SECONDS_TO_MILLISECONDS,
    verboseLevel: 3,
    esm: lambdaPath.endsWith('.mjs'),
  })

  return result
}
