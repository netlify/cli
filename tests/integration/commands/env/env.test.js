import process from 'process'

import execa from 'execa'
import { describe, test } from 'vitest'

import { callCli } from '../../utils/call-cli.js'
import { cliPath } from '../../utils/cli-path.js'
import { CONFIRM, answerWithValue, handleQuestions } from '../../utils/handle-questions.js'
import { getCLIOptions, withMockApi } from '../../utils/mock-api.js'
import { withSiteBuilder } from '../../utils/site-builder.ts'
import { normalize } from '../../utils/snapshots.js'

const siteInfo = {
  account_slug: 'test-account',
  build_settings: { env: {} },
  id: 'site_id',
  name: 'site-name',
}
const existingVar = {
  key: 'EXISTING_VAR',
  scopes: ['builds', 'functions'],
  values: [
    {
      id: '1234',
      context: 'production',
      value: 'prod-value',
    },
    {
      id: '2345',
      context: 'dev',
      value: 'dev-value',
    },
  ],
}
const otherVar = {
  key: 'OTHER_VAR',
  scopes: ['builds', 'functions', 'runtime', 'post_processing'],
  values: [
    {
      id: '3456',
      context: 'all',
      value: 'all-value',
    },
  ],
}
const routes = [
  { path: 'sites/site_id', response: siteInfo },
  { path: 'sites/site_id/service-instances', response: [] },
  { path: 'accounts', response: [{ slug: siteInfo.account_slug }] },
]

describe('commands/env', () => {
  describe('env:get', () => {
    test('--json should return empty object if var not set', async (t) => {
      await withSiteBuilder(t, async (builder) => {
        await builder.build()

        await withMockApi(routes, async ({ apiUrl }) => {
          const cliResponse = await callCli(['env:get', '--json', 'SOME_VAR'], getCLIOptions({ builder, apiUrl }), true)

          t.expect(cliResponse).toStrictEqual({})
        })
      })
    })

    test('--json should return value of existing var', async (t) => {
      await withSiteBuilder(t, async (builder) => {
        await builder.build()

        const getRoutes = [...routes, { path: 'accounts/test-account/env/EXISTING_VAR', response: existingVar }]

        await withMockApi(getRoutes, async ({ apiUrl }) => {
          const cliResponse = await callCli(
            ['env:get', '--json', 'EXISTING_VAR'],
            getCLIOptions({ builder, apiUrl }),
            true,
          )

          t.expect(cliResponse).toStrictEqual({ EXISTING_VAR: 'dev-value' })
        })
      })
    })

    test('--json should return value of var from netlify.toml', async (t) => {
      await withSiteBuilder(t, async (builder) => {
        await builder
          .withNetlifyToml({
            config: {
              build: {
                environment: { FROM_TOML: 'FROM_TOML' },
              },
            },
          })
          .build()

        await withMockApi(routes, async ({ apiUrl }) => {
          const cliResponse = await callCli(
            ['env:get', '--json', 'FROM_TOML'],
            getCLIOptions({ builder, apiUrl }),
            true,
          )

          t.expect(cliResponse).toStrictEqual({ FROM_TOML: 'FROM_TOML' })
        })
      })
    })
  })

  describe('env:list', () => {
    test('--json should return empty object if no vars set', async (t) => {
      await withSiteBuilder(t, async (builder) => {
        await builder.build()

        const listRoutes = [...routes, { path: 'accounts/test-account/env', response: [] }]

        await withMockApi(listRoutes, async ({ apiUrl }) => {
          const cliResponse = await callCli(['env:list', '--json'], getCLIOptions({ builder, apiUrl }), true)

          t.expect(cliResponse).toStrictEqual({})
        })
      })
    })

    test('--json should return list of vars with netlify.toml taking priority', async (t) => {
      await withSiteBuilder(t, async (builder) => {
        await builder
          .withNetlifyToml({
            config: {
              build: {
                environment: { EXISTING_VAR: 'TOML_VALUE' },
              },
            },
          })
          .build()

        const listRoutes = [...routes, { path: 'accounts/test-account/env', response: [existingVar, otherVar] }]

        await withMockApi(listRoutes, async ({ apiUrl }) => {
          const cliResponse = await callCli(['env:list', '--json'], getCLIOptions({ builder, apiUrl }), true)

          t.expect(cliResponse).toStrictEqual({
            EXISTING_VAR: 'TOML_VALUE',
            OTHER_VAR: 'all-value',
          })
        })
      })
    })

    test('should hide variables values and prompt to show', async (t) => {
      await withSiteBuilder(t, async (builder) => {
        await builder.build()

        const questions = [
          {
            question: 'Show values',
            answer: CONFIRM,
          },
        ]

        const listRoutes = [...routes, { path: 'accounts/test-account/env', response: [existingVar, otherVar] }]

        await withMockApi(listRoutes, async ({ apiUrl }) => {
          // we set extendEnv: false to prevent the CLI detecting GitHub Actions as CI
          const childProcess = execa(
            cliPath,
            ['env:list'],
            getCLIOptions({
              apiUrl,
              builder,
              extendEnv: false,
              env: { PATH: process.env.PATH, HOME: process.env.HOME, APPDATA: process.env.APPDATA },
            }),
          )

          handleQuestions(childProcess, questions)

          const { stdout: cliResponse } = await childProcess

          t.expect(normalize(cliResponse)).toMatchSnapshot()
        })
      })
    })

    test('should hide variables values and show on confirm', async (t) => {
      await withSiteBuilder(t, async (builder) => {
        await builder.build()

        const questions = [
          {
            question: 'Show values',
            answer: answerWithValue('y'),
          },
        ]

        const listRoutes = [...routes, { path: 'accounts/test-account/env', response: [existingVar, otherVar] }]

        await withMockApi(listRoutes, async ({ apiUrl }) => {
          // we set extendEnv: false to prevent the CLI detecting GitHub Actions as CI
          const childProcess = execa(
            cliPath,
            ['env:list'],
            getCLIOptions({
              apiUrl,
              builder,
              extendEnv: false,
              env: { PATH: process.env.PATH, HOME: process.env.HOME, APPDATA: process.env.APPDATA },
            }),
          )

          handleQuestions(childProcess, questions)

          const { stdout: cliResponse } = await childProcess

          t.expect(normalize(cliResponse)).toMatchSnapshot()
        })
      })
    })

    test('should not prompt on CI', async (t) => {
      await withSiteBuilder(t, async (builder) => {
        await builder.build()

        const listRoutes = [...routes, { path: 'accounts/test-account/env', response: [existingVar, otherVar] }]

        await withMockApi(listRoutes, async ({ apiUrl }) => {
          const cliResponse = await callCli(['env:list'], getCLIOptions({ builder, apiUrl, env: { CI: true } }))

          t.expect(normalize(cliResponse)).toMatchSnapshot()
        })
      })
    })
  })

  describe('env:set', () => {
    test('--json should create and return new var', async (t) => {
      await withSiteBuilder(t, async (builder) => {
        await builder.build()

        const setRoutes = [
          ...routes,
          { path: 'accounts/test-account/env', response: [existingVar, otherVar] },
          { path: 'accounts/test-account/env', method: 'POST', response: {} },
        ]

        await withMockApi(setRoutes, async ({ apiUrl }) => {
          const cliResponse = await callCli(
            ['env:set', '--json', 'NEW_VAR', 'new-value', '--force'],
            getCLIOptions({ builder, apiUrl }),
            true,
          )

          t.expect(cliResponse).toStrictEqual({
            EXISTING_VAR: 'dev-value',
            OTHER_VAR: 'all-value',
            NEW_VAR: 'new-value',
          })
        })
      })
    })

    test('--json should update existing var', async (t) => {
      await withSiteBuilder(t, async (builder) => {
        await builder.build()

        const setRoutes = [
          ...routes,
          { path: 'accounts/test-account/env', response: [existingVar, otherVar] },
          { path: 'accounts/test-account/env/EXISTING_VAR', method: 'PUT', response: {} },
        ]

        await withMockApi(setRoutes, async ({ apiUrl }) => {
          const cliResponse = await callCli(
            ['env:set', '--json', 'EXISTING_VAR', 'new-value', '--force'],
            getCLIOptions({ builder, apiUrl }),
            true,
          )

          t.expect(cliResponse).toStrictEqual({
            EXISTING_VAR: 'new-value',
            OTHER_VAR: 'all-value',
          })
        })
      })
    })
  })

  describe('env:unset', () => {
    test('--json should remove existing variable', async (t) => {
      await withSiteBuilder(t, async (builder) => {
        await builder.build()

        const unsetRoutes = [
          ...routes,
          { path: 'accounts/test-account/env', response: [existingVar, otherVar] },
          { path: 'accounts/test-account/env/EXISTING_VAR', method: 'DELETE', response: {} },
        ]

        await withMockApi(unsetRoutes, async ({ apiUrl }) => {
          const cliResponse = await callCli(
            ['env:unset', '--json', 'EXISTING_VAR', '--force'],
            getCLIOptions({ builder, apiUrl }),
            true,
          )

          t.expect(cliResponse).toStrictEqual({
            OTHER_VAR: 'all-value',
          })
        })
      })
    })
  })

  describe('env:import', () => {
    test('env:import should throw error if file not exists', async (t) => {
      await withSiteBuilder(t, async (builder) => {
        await builder.build()

        await withMockApi(routes, async ({ apiUrl }) => {
          await t.expect(() => callCli(['env:import', '.env'], getCLIOptions({ builder, apiUrl }))).rejects.toThrow()
        })
      })
    })

    test('--json should import new vars and override existing vars', async (t) => {
      await withSiteBuilder(t, async (builder) => {
        await builder
          .withEnvFile({
            path: '.env',
            env: {
              EXISTING_VAR: 'new-value',
              NEW_VAR: 'new-value',
            },
          })
          .build()

        const importRoutes = [
          ...routes,
          { path: 'accounts/test-account/env', response: [existingVar] },
          { path: 'accounts/test-account/env', method: 'POST', response: {} },
          { path: 'accounts/test-account/env/EXISTING_VAR', method: 'DELETE', response: {} },
        ]

        await withMockApi(importRoutes, async ({ apiUrl }) => {
          const cliResponse = await callCli(['env:import', '--json', '.env'], getCLIOptions({ builder, apiUrl }), true)

          t.expect(cliResponse).toStrictEqual({
            EXISTING_VAR: 'new-value',
            NEW_VAR: 'new-value',
          })
        })
      })
    })

    test('--json --replace-existing should replace all existing vars and return imported', async (t) => {
      await withSiteBuilder(t, async (builder) => {
        await builder
          .withEnvFile({
            path: '.env',
            env: {
              NEW_VAR: 'new-value',
            },
          })
          .build()

        const importRoutes = [
          ...routes,
          { path: 'accounts/test-account/env', response: [existingVar] },
          { path: 'accounts/test-account/env', method: 'POST', response: [] },
          { path: 'accounts/test-account/env/EXISTING_VAR', method: 'DELETE', response: {} },
        ]

        await withMockApi(importRoutes, async ({ apiUrl }) => {
          const cliResponse = await callCli(
            ['env:import', '--replace-existing', '--json', '.env'],
            getCLIOptions({ builder, apiUrl }),
            true,
          )

          t.expect(cliResponse).toStrictEqual({
            NEW_VAR: 'new-value',
          })
        })
      })
    })
  })

  describe('env:clone', () => {
    test("should return without clone if there's no env in source site", async (t) => {
      await withSiteBuilder(t, async (builder) => {
        await builder.build()
        const createRoutes = [
          { path: 'accounts', response: [{ slug: siteInfo.account_slug }] },
          { path: 'accounts/test-account/env', response: [] },
          { path: 'sites/site_id/service-instances', response: [] },
          { path: 'sites/site_id', response: { ...siteInfo, build_settings: { env: {} } } },
          { path: 'sites/site_id_a', response: { ...siteInfo, build_settings: { env: {} } } },
        ]
        await withMockApi(createRoutes, async ({ apiUrl }) => {
          const cliResponse = await callCli(
            ['env:clone', '--to', 'site_id_a', '--force'],
            getCLIOptions({ builder, apiUrl }),
          )

          t.expect(normalize(cliResponse)).toMatchSnapshot()
        })
      })
    })

    test("should print error if --to site doesn't exist", async (t) => {
      await withSiteBuilder(t, async (builder) => {
        await builder.build()
        const createRoutes = [{ path: 'sites/site_id', response: { ...siteInfo, build_settings: { env: {} } } }]
        await withMockApi(createRoutes, async ({ apiUrl }) => {
          const { stderr: cliResponse } = await callCli(
            ['env:clone', '--to', 'to-site', '--force'],
            getCLIOptions({ builder, apiUrl }),
          ).catch((error) => error)

          t.expect(cliResponse.includes(`Can't find site with id to-site. Please make sure the site exists`)).toBe(true)
        })
      })
    })

    test("should print error if --from site doesn't exist", async (t) => {
      await withSiteBuilder(t, async (builder) => {
        await builder.build()
        await withMockApi([], async ({ apiUrl }) => {
          const { stderr: cliResponse } = await callCli(
            ['env:clone', '--from', 'from-site', '--to', 'to-site', '--force'],
            getCLIOptions({ builder, apiUrl }),
          ).catch((error) => error)

          t.expect(cliResponse.includes(`Can't find site with id from-site. Please make sure the site exists`)).toBe(
            true,
          )
        })
      })
    })

    test('should exit if the folder is not linked to a site, and --from is not provided', async (t) => {
      await withSiteBuilder(t, async (builder) => {
        await builder.build()

        const cliResponse = await callCli(['env:clone', '--to', 'site_id_a', '--force'], {
          cwd: builder.directory,
          extendEnv: false,
          PATH: process.env.PATH,
        })

        t.expect(normalize(cliResponse)).toMatchSnapshot()
      })
    })

    test('should return success message', async (t) => {
      const cloneRoutes = [
        { path: 'accounts', response: [{ slug: siteInfo.account_slug }] },
        { path: 'sites/site_id/service-instances', response: [] },
        { path: 'sites/site_id', response: siteInfo },
        {
          path: 'sites/site_id_a',
          response: {
            id: 'site_id_a',
            name: 'site-name-a',
            account_slug: 'test-account-a',
            build_settings: { env: {} },
          },
        },
        { path: 'accounts/test-account/env', response: [existingVar] },
        { path: 'accounts/test-account-a/env', response: [otherVar] },
        { path: 'accounts/test-account-a/env', method: 'POST', response: {} },
      ]

      await withSiteBuilder(t, async (builder) => {
        await builder.build()
        await withMockApi(cloneRoutes, async ({ apiUrl, requests }) => {
          const cliResponse = await callCli(
            ['env:clone', '--to', 'site_id_a', '--force'],
            getCLIOptions({ apiUrl, builder }),
          )

          t.expect(normalize(cliResponse)).toMatchSnapshot()

          const postRequest = requests.find((request) => request.method === 'POST')
          t.expect(postRequest.body).toStrictEqual([existingVar])
        })
      })
    })
  })
})
