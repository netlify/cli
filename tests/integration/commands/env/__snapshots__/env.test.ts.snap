// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`commands/env > env:clone > should exit if the folder is not linked to a site, and --from is not provided 1`] = `"Please include the source site Id as the \`--from\` option, or run \`netlify link\` to link this folder to a Netlify site"`;

exports[`commands/env > env:clone > should return success message 1`] = `"Successfully cloned environment variables from site-name to site-name-a"`;

exports[`commands/env > env:clone > should return without clone if there's no env in source site 1`] = `"site-name has no environment variables, nothing to clone"`;

exports[`commands/env > env:list > should hide variables values and prompt to show 1`] = `
"2 environment variables for site site-name in the dev context 
.-------------------------------------------------------------------------------
--------.
|                                 Environment variables                         
        |
|-------------------------------------------------------------------------------
--------|
|     Key      |                       Value                        |       Scop
e       |
|--------------|----------------------------------------------------|-----------
--------|
| EXISTING_VAR | ************************************************** | Builds, Fu
nctions |
| OTHER_VAR    | ************************************************** | All       
        |
'-------------------------------------------------------------------------------
--------'
? Show values? (y/N) [21D[21C[2K[G? Show values? No[17D[17C"
`;

exports[`commands/env > env:list > should hide variables values and show on confirm 1`] = `
"2 environment variables for site site-name in the dev context 
.-------------------------------------------------------------------------------
--------.
|                                 Environment variables                         
        |
|-------------------------------------------------------------------------------
--------|
|     Key      |                       Value                        |       Scop
e       |
|--------------|----------------------------------------------------|-----------
--------|
| EXISTING_VAR | ************************************************** | Builds, Fu
nctions |
| OTHER_VAR    | ************************************************** | All       
        |
'-------------------------------------------------------------------------------
--------'
? Show values? (y/N) [21D[21C[2K[G? Show values? (y/N) y[22D[22C[2K[G? Show values? Yes[18D[18C
[2K[1A[2K[1A[2K[G
[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[1A[2K[G.----------------------------------------------.
|            Environment variables             |
|----------------------------------------------|
|     Key      |   Value   |       Scope       |
|--------------|-----------|-------------------|
| EXISTING_VAR | dev-value | Builds, Functions |
| OTHER_VAR    | all-value | All               |
'----------------------------------------------'
? Show values? Yes"
`;

exports[`commands/env > env:list > should not prompt on CI 1`] = `
"2 environment variables for site site-name in the dev context 
.----------------------------------------------.
|            Environment variables             |
|----------------------------------------------|
|     Key      |   Value   |       Scope       |
|--------------|-----------|-------------------|
| EXISTING_VAR | dev-value | Builds, Functions |
| OTHER_VAR    | all-value | All               |
'----------------------------------------------'"
`;
